File: src/http_client.c
/*
 * http_client.c - Unified HTTP client implementation
 */

#define _POSIX_C_SOURCE 200809L

#include "http_client.h"
#include "logger.h"
#include "retry_logic.h"  // For common retry logic
#include "sse_parser.h"   // SSE parser for streaming

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <errno.h>

// ============================================================================
// Internal Types
// ============================================================================

/* HTTP response buffer - separate from SSE parser's HttpResponseBuffer */
typedef struct {
    char *data;
    size_t size;
    size_t capacity;
} HttpResponseBuffer;

typedef struct {
    HttpResponseBuffer *buffer;
    struct curl_slist *headers;
} WriteContext;

// ============================================================================
// Static Helpers
// ============================================================================

static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    size_t realsize = size * nmemb;
    WriteContext *ctx = (WriteContext *)userp;
    HttpResponseBuffer *buf = ctx->buffer;

    // Ensure we have enough capacity
    size_t needed = buf->size + realsize + 1;
    if (needed > buf->capacity) {
        // Grow by at least 2x or to needed size
        size_t new_capacity = buf->capacity * 2;
        if (new_capacity < needed) {
            new_capacity = needed;
        }

        char *new_data = realloc(buf->data, new_capacity);
        if (!new_data) {
            LOG_ERROR("Failed to allocate memory for HTTP response (needed: %zu)", needed);
            return 0;
        }

        buf->data = new_data;
        buf->capacity = new_capacity;
    }

    // Copy data
    memcpy(buf->data + buf->size, contents, realsize);
    buf->size += realsize;
    buf->data[buf->size] = '\0';

    return realsize;
}

static size_t header_callback(char *buffer, size_t size, size_t nitems, void *userdata) {
    size_t realsize = size * nitems;
    WriteContext *ctx = (WriteContext *)userdata;

    // Null-terminate the header line
    char *header = malloc(realsize + 1);
    if (!header) {
        LOG_ERROR("Failed to allocate memory for header");
        return 0;
    }

    memcpy(header, buffer, realsize);
    header[realsize] = '\0';

    // Remove trailing CRLF
    if (realsize >= 2 && header[realsize-2] == '\r' && header[realsize-1] == '\n') {
        header[realsize-2] = '\0';
    } else if (realsize >= 1 && header[realsize-1] == '\n') {
        header[realsize-1] = '\0';
    }

    // Add to headers list
    ctx->headers = curl_slist_append(ctx->headers, header);
    free(header);

    return realsize;
}

static HttpResponseBuffer* http_response_buffer_create(void) {
    HttpResponseBuffer *buf = calloc(1, sizeof(HttpResponseBuffer));
    if (!buf) {
        return NULL;
    }

    // Initial capacity
    buf->capacity = 4096;
    buf->data = malloc(buf->capacity);
    if (!buf->data) {
        free(buf);
        return NULL;
    }

    buf->data[0] = '\0';
    return buf;
}

static void http_response_buffer_free(HttpResponseBuffer *buf) {
    if (buf) {
        free(buf->data);
        free(buf);
    }
}

// Callback for streaming data (uses SSE parser)
static size_t streaming_write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    size_t realsize = size * nmemb;
    SSEParserState *parser = (SSEParserState *)userp;

    if (parser->abort_requested) {
        return 0;  // Abort the transfer
    }

    // Use the SSE parser to process the data
    int result = sse_parser_process_data(parser, (const char *)contents, realsize);
    if (result != 0) {
        return 0;  // Abort
    }

    return realsize;
}



// ============================================================================
// Public Functions
// ============================================================================

int http_client_init(void) {
    CURLcode res = curl_global_init(CURL_GLOBAL_DEFAULT);
    if (res != CURLE_OK) {
        LOG_ERROR("Failed to initialize libcurl: %s", curl_easy_strerror(res));
        return -1;
    }
    return 0;
}

void http_client_cleanup(void) {
    curl_global_cleanup();
}

HttpResponse* http_client_execute(const HttpRequest *req,
                                 HttpProgressCallback progress_cb,
                                 void *progress_data) {
    if (!req || !req->url) {
        LOG_ERROR("Invalid HTTP request: NULL request or URL");
        return NULL;
    }

    HttpResponse *resp = calloc(1, sizeof(HttpResponse));
    if (!resp) {
        LOG_ERROR("Failed to allocate HTTP response");
        return NULL;
    }

    CURL *curl = curl_easy_init();
    if (!curl) {
        LOG_ERROR("Failed to initialize CURL handle");
        free(resp);
        return NULL;
    }

    // Initialize buffers
    HttpResponseBuffer *body_buf = http_response_buffer_create();
    WriteContext write_ctx = {0};
    write_ctx.buffer = body_buf;

    if (!body_buf) {
        LOG_ERROR("Failed to allocate memory buffer");
        curl_easy_cleanup(curl);
        free(resp);
        return NULL;
    }

    // Set up curl options
    curl_easy_setopt(curl, CURLOPT_URL, req->url);

    // Method
    if (req->method && strcmp(req->method, "GET") == 0) {
        curl_easy_setopt(curl, CURLOPT_HTTPGET, 1L);
    } else if (req->method && strcmp(req->method, "POST") == 0) {
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        if (req->body) {
            curl_easy_setopt(curl, CURLOPT_POSTFIELDS, req->body);
        }
    } else if (req->method && strcmp(req->method, "PUT") == 0) {
        curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PUT");
        if (req->body) {
            curl_easy_setopt(curl, CURLOPT_POSTFIELDS, req->body);
        }
    } else {
        // Default to POST
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        if (req->body) {
            curl_easy_setopt(curl, CURLOPT_POSTFIELDS, req->body);
        }
    }

    // Headers
    if (req->headers) {
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, req->headers);
    }

    // Timeouts (convert ms to seconds for curl)
    if (req->connect_timeout_ms > 0) {
        curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT_MS, req->connect_timeout_ms);
    } else {
        curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 30L); // Default 30 seconds
    }

    if (req->total_timeout_ms > 0) {
        curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, req->total_timeout_ms);
    } else {
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 300L); // Default 5 minutes
    }

    // Redirects
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, req->follow_redirects ? 1L : 0L);

    // Verbose logging
    curl_easy_setopt(curl, CURLOPT_VERBOSE, req->verbose ? 1L : 0L);

    // Callbacks
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &write_ctx);

    curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, header_callback);
    curl_easy_setopt(curl, CURLOPT_HEADERDATA, &write_ctx);

    // Progress callback
    if (progress_cb) {
        curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 0L);
        curl_easy_setopt(curl, CURLOPT_XFERINFOFUNCTION, progress_cb);
        curl_easy_setopt(curl, CURLOPT_XFERINFODATA, progress_data);
    } else {
        curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 1L);
    }

    // Execute request
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    CURLcode res = curl_easy_perform(curl);

    clock_gettime(CLOCK_MONOTONIC, &end);
    resp->duration_ms = (end.tv_sec - start.tv_sec) * 1000 +
                       (end.tv_nsec - start.tv_nsec) / 1000000;

    // Get status code
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &resp->status_code);

    // Handle result
    if (res != CURLE_OK) {
        // Check if the error was due to user interruption
        if (res == CURLE_ABORTED_BY_CALLBACK) {
            resp->error_message = strdup("Request interrupted by user");
            resp->is_retryable = 0;
        } else {
            resp->error_message = strdup(curl_easy_strerror(res));
            // Network/protocol errors that should be retried:
            // - Connection issues (COULDNT_CONNECT, RECV_ERROR, SEND_ERROR, GOT_NOTHING)
            // - Timeouts (OPERATION_TIMEDOUT)
            // - SSL issues (SSL_CONNECT_ERROR)
            // - HTTP2/HTTP3 protocol layer issues (HTTP2, HTTP2_STREAM)
            resp->is_retryable = is_curl_error_retryable(res);
        }

        // Clean up buffers on error
        http_response_buffer_free(body_buf);
        curl_slist_free_all(write_ctx.headers);
    } else {
        // Success - transfer ownership
        resp->body = body_buf->data;
        body_buf->data = NULL; // Prevent double free
        resp->headers = write_ctx.headers;

        // Body buffer is now owned by HttpResponse
        free(body_buf);
    }

    curl_easy_cleanup(curl);
    return resp;
}

void http_response_free(HttpResponse *resp) {
    if (!resp) return;

    free(resp->body);
    curl_slist_free_all(resp->headers);
    free(resp->error_message);
    free(resp);
}

char* http_headers_to_json(struct curl_slist *headers) {
    if (!headers) {
        return NULL;
    }

    cJSON *headers_array = cJSON_CreateArray();
    if (!headers_array) {
        return NULL;
    }

    struct curl_slist *current = headers;
    while (current) {
        if (current->data) {
            cJSON *header_obj = cJSON_CreateObject();
            if (header_obj) {
                // Parse header line into name and value
                char *colon = strchr(current->data, ':');
                if (colon) {
                    *colon = '\0';  // Split the string
                    char *header_name = current->data;
                    char *header_value = colon + 1;

                    // Skip leading whitespace in value
                    while (*header_value == ' ' || *header_value == '\t') {
                        header_value++;
                    }

                    cJSON_AddStringToObject(header_obj, "name", header_name);
                    cJSON_AddStringToObject(header_obj, "value", header_value);

                    *colon = ':';  // Restore the colon
                } else {
                    // If no colon, treat the whole line as a header line
                    cJSON_AddStringToObject(header_obj, "line", current->data);
                }
                cJSON_AddItemToArray(headers_array, header_obj);
            }
        }
        current = current->next;
    }

    char *json_str = cJSON_PrintUnformatted(headers_array);
    cJSON_Delete(headers_array);
    return json_str;
}

struct curl_slist* http_headers_deep_copy(const struct curl_slist *headers) {
    if (!headers) {
        return NULL;
    }

    struct curl_slist *new_list = NULL;
    const struct curl_slist *current = headers;

    while (current) {
        if (current->data) {
            struct curl_slist *new_item = curl_slist_append(new_list, current->data);
            if (!new_item) {
                curl_slist_free_all(new_list);
                return NULL;
            }
            new_list = new_item;
        }
        current = current->next;
    }

    return new_list;
}

struct curl_slist* http_copy_headers(struct curl_slist *headers) {
    struct curl_slist *new_list = NULL;
    struct curl_slist *current = headers;

    while (current) {
        if (current->data) {
            new_list = curl_slist_append(new_list, current->data);
            if (!new_list) {
                // Failed to allocate, clean up what we've created
                curl_slist_free_all(new_list);
                return NULL;
            }
        }
        current = current->next;
    }

    return new_list;
}

struct curl_slist* http_add_header(struct curl_slist *headers, const char *header) {
    return curl_slist_append(headers, header);
}

HttpResponse* http_client_execute_stream(const HttpRequest *req,
                                        HttpStreamCallback stream_cb,
                                        void *stream_data,
                                        HttpProgressCallback progress_cb,
                                        void *progress_data) {
    if (!req || !req->url || !stream_cb) {
        LOG_ERROR("Invalid streaming HTTP request: NULL request, URL, or stream callback");
        return NULL;
    }

    HttpResponse *resp = calloc(1, sizeof(HttpResponse));
    if (!resp) {
        LOG_ERROR("Failed to allocate HTTP response");
        return NULL;
    }

    CURL *curl = curl_easy_init();
    if (!curl) {
        LOG_ERROR("Failed to initialize CURL handle");
        free(resp);
        return NULL;
    }

    // Initialize SSE parser
    SSEParserState *parser = sse_parser_create(stream_cb, stream_data);
    if (!parser) {
        LOG_ERROR("Failed to allocate SSE parser");
        curl_easy_cleanup(curl);
        free(resp);
        return NULL;
    }

    // Set up headers capture
    WriteContext write_ctx = {0};
    write_ctx.buffer = NULL;  // No body buffer for streaming

    // Set up curl options
    curl_easy_setopt(curl, CURLOPT_URL, req->url);

    // Method
    if (req->method && strcmp(req->method, "GET") == 0) {
        curl_easy_setopt(curl, CURLOPT_HTTPGET, 1L);
    } else if (req->method && strcmp(req->method, "POST") == 0) {
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        if (req->body) {
            curl_easy_setopt(curl, CURLOPT_POSTFIELDS, req->body);
        }
    } else if (req->method && strcmp(req->method, "PUT") == 0) {
        curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PUT");
        if (req->body) {
            curl_easy_setopt(curl, CURLOPT_POSTFIELDS, req->body);
        }
    } else {
        // Default to POST
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        if (req->body) {
            curl_easy_setopt(curl, CURLOPT_POSTFIELDS, req->body);
        }
    }

    // Headers
    if (req->headers) {
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, req->headers);
    }

    // Timeouts
    if (req->connect_timeout_ms > 0) {
        curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT_MS, req->connect_timeout_ms);
    } else {
        curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 30L);
    }

    if (req->total_timeout_ms > 0) {
        curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, req->total_timeout_ms);
    } else {
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 300L);
    }

    // Redirects
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, req->follow_redirects ? 1L : 0L);

    // Verbose logging
    curl_easy_setopt(curl, CURLOPT_VERBOSE, req->verbose ? 1L : 0L);

    // Streaming write callback
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, streaming_write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, parser);

    // Headers callback
    curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, header_callback);
    curl_easy_setopt(curl, CURLOPT_HEADERDATA, &write_ctx);

    // Progress callback
    if (progress_cb) {
        curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 0L);
        curl_easy_setopt(curl, CURLOPT_XFERINFOFUNCTION, progress_cb);
        curl_easy_setopt(curl, CURLOPT_XFERINFODATA, progress_data);
    } else {
        curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 1L);
    }

    // Execute request
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    CURLcode res = curl_easy_perform(curl);

    clock_gettime(CLOCK_MONOTONIC, &end);
    resp->duration_ms = (end.tv_sec - start.tv_sec) * 1000 +
                       (end.tv_nsec - start.tv_nsec) / 1000000;

    // Get status code
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &resp->status_code);

    // Handle result
    if (res != CURLE_OK) {
        if (res == CURLE_ABORTED_BY_CALLBACK) {
            resp->error_message = strdup(parser->abort_requested ?
                                        "Stream aborted by callback" :
                                        "Request interrupted by user");
            resp->is_retryable = 0;
        } else {
            resp->error_message = strdup(curl_easy_strerror(res));
            resp->is_retryable = is_curl_error_retryable(res);
        }
    }

    // Transfer headers
    resp->headers = write_ctx.headers;

    // Cleanup
    sse_parser_free(parser);
    curl_easy_cleanup(curl);

    return resp;
}

File: zig/http_client.zig
//! http_client.zig — HTTP client wrapping libcurl via @cImport
//!
//! Zig port of src/http_client.c / src/http_client.h.
//!
//! Keeps libcurl as a C library (via `@cImport`) and wraps the call sites with
//! Zig error unions so callers never touch raw curl error codes.
//!
//! ## Public surface
//! - `HttpClient` — owns a `CURL*` handle; call `init()`/`deinit()`
//! - `Request` — URL, method, headers, body, timeouts
//! - `Response` — status code + owned body bytes; call `deinit(allocator)`
//! - `streamRequest()` — SSE streaming variant; calls `StreamCallback` per line
//! - `globalInit()` / `globalCleanup()` — must wrap program lifetime

const std = @import("std");

const c = @cImport({
    @cInclude("curl/curl.h");
});

// ---------------------------------------------------------------------------
// Error set
// ---------------------------------------------------------------------------

pub const HttpError = error{
    /// `curl_global_init` failed.
    CurlGlobalInitFailed,
    /// `curl_easy_init` returned NULL.
    CurlInitFailed,
    /// A curl perform error occurred (non-abort).
    CurlError,
    /// The stream callback returned non-zero to abort.
    StreamAborted,
    /// Memory allocation failed.
    OutOfMemory,
};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

pub const Method = enum {
    GET,
    POST,
    PUT,
    DELETE,
    PATCH,
};

pub const Header = struct {
    name: []const u8,
    value: []const u8,
};

pub const Request = struct {
    url: []const u8,
    method: Method = .POST,
    /// Slice of `Header` values; the client copies them into curl slist.
    headers: []const Header = &.{},
    /// Optional request body.
    body: ?[]const u8 = null,
    /// Connection timeout in milliseconds (0 = curl default 30 s).
    connect_timeout_ms: u32 = 30_000,
    /// Total transfer timeout in milliseconds (0 = curl default 5 min).
    total_timeout_ms: u32 = 300_000,
    /// Follow HTTP redirects.
    follow_redirects: bool = false,
    /// Enable curl verbose logging (useful for debugging).
    verbose: bool = false,
};

pub const Response = struct {
    status_code: u32,
    /// Owned bytes — caller must call `deinit(allocator)`.
    body: []u8,
    /// Request duration in milliseconds.
    duration_ms: i64,

    pub fn deinit(self: *Response, allocator: std.mem.Allocator) void {
        allocator.free(self.body);
    }
};

/// Callback type for SSE streaming.
/// `line` is a single SSE text line (not including the trailing newline).
/// Return `false` to continue streaming; `true` to abort.
pub const StreamCallback = *const fn (line: []const u8, userdata: ?*anyopaque) bool;

// ---------------------------------------------------------------------------
// Internal write-buffer context
// ---------------------------------------------------------------------------

const WriteCtx = struct {
    buf: std.ArrayList(u8),
};

fn writeCallback(
    data: [*]u8,
    size: usize,
    nmemb: usize,
    userdata: *anyopaque,
) callconv(.C) usize {
    const ctx = @as(*WriteCtx, @ptrCast(@alignCast(userdata)));
    const bytes = data[0 .. size * nmemb];
    ctx.buf.appendSlice(bytes) catch return 0;
    return size * nmemb;
}

// ---------------------------------------------------------------------------
// SSE streaming context
// ---------------------------------------------------------------------------

const StreamCtx = struct {
    callback: StreamCallback,
    userdata: ?*anyopaque,
    abort_requested: bool,
    /// Leftover bytes from a previous chunk that didn't end on a newline.
    partial: std.ArrayList(u8),
};

fn streamWriteCallback(
    data: [*]u8,
    size: usize,
    nmemb: usize,
    userdata: *anyopaque,
) callconv(.C) usize {
    const realsize = size * nmemb;
    const ctx = @as(*StreamCtx, @ptrCast(@alignCast(userdata)));

    if (ctx.abort_requested) return 0;

    const chunk = data[0..realsize];
    var pos: usize = 0;

    while (pos < chunk.len) {
        // Find next newline in the incoming chunk
        const eol = std.mem.indexOfScalar(u8, chunk[pos..], '\n');
        if (eol == null) {
            // No newline — buffer the rest for the next callback invocation
            ctx.partial.appendSlice(chunk[pos..]) catch {
                ctx.abort_requested = true;
                return 0;
            };
            break;
        }

        const nl_pos = pos + eol.?;
        // Assemble the line: partial + chunk up to (not including) newline
        ctx.partial.appendSlice(chunk[pos..nl_pos]) catch {
            ctx.abort_requested = true;
            return 0;
        };

        // Strip trailing \r if present
        var line = ctx.partial.items;
        if (line.len > 0 and line[line.len - 1] == '\r') {
            line = line[0 .. line.len - 1];
        }

        // Dispatch to callback
        if (ctx.callback(line, ctx.userdata)) {
            ctx.abort_requested = true;
            return 0;
        }

        ctx.partial.clearRetainingCapacity();
        pos = nl_pos + 1; // skip past \n
    }

    return realsize;
}

// ---------------------------------------------------------------------------
// Global init / cleanup
// ---------------------------------------------------------------------------

/// Must be called once before any HTTP requests (typically at program start).
pub fn globalInit() HttpError!void {
    const rc = c.curl_global_init(c.CURL_GLOBAL_DEFAULT);
    if (rc != c.CURLE_OK) return error.CurlGlobalInitFailed;
}

/// Must be called once at program exit.
pub fn globalCleanup() void {
    c.curl_global_cleanup();
}

// ---------------------------------------------------------------------------
// HttpClient
// ---------------------------------------------------------------------------

pub const HttpClient = struct {
    handle: *c.CURL,

    /// Create a new client.  The internal `CURL*` handle is reused across
    /// calls; each `request()` call resets options via `curl_easy_reset`.
    pub fn init() HttpError!HttpClient {
        const handle = c.curl_easy_init() orelse return error.CurlInitFailed;
        return HttpClient{ .handle = handle };
    }

    pub fn deinit(self: *HttpClient) void {
        c.curl_easy_cleanup(self.handle);
    }

    // ------------------------------------------------------------------
    // Non-streaming request
    // ------------------------------------------------------------------

    /// Perform a blocking HTTP request and return the full response body.
    /// The returned `Response.body` is owned by the caller (allocated with
    /// `allocator`); call `response.deinit(allocator)` to free it.
    pub fn request(
        self: *HttpClient,
        allocator: std.mem.Allocator,
        req: Request,
    ) (HttpError || std.mem.Allocator.Error)!Response {
        c.curl_easy_reset(self.handle);

        var write_ctx = WriteCtx{ .buf = std.ArrayList(u8).init(allocator) };
        defer write_ctx.buf.deinit();

        // Build null-terminated URL
        const url_z = try std.fmt.allocPrintZ(allocator, "{s}", .{req.url});
        defer allocator.free(url_z);

        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_URL, url_z.ptr);

        // Method + body
        try setMethod(self.handle, req.method, req.body);

        // Headers
        var slist: ?*c.curl_slist = null;
        defer if (slist) |sl| c.curl_slist_free_all(sl);
        for (req.headers) |h| {
            const header_str = try std.fmt.allocPrintZ(allocator, "{s}: {s}", .{ h.name, h.value });
            defer allocator.free(header_str);
            slist = c.curl_slist_append(slist, header_str.ptr);
        }
        if (slist) |sl| _ = c.curl_easy_setopt(self.handle, c.CURLOPT_HTTPHEADER, sl);

        // Timeouts
        if (req.connect_timeout_ms > 0) {
            _ = c.curl_easy_setopt(self.handle, c.CURLOPT_CONNECTTIMEOUT_MS, @as(c_long, req.connect_timeout_ms));
        }
        if (req.total_timeout_ms > 0) {
            _ = c.curl_easy_setopt(self.handle, c.CURLOPT_TIMEOUT_MS, @as(c_long, req.total_timeout_ms));
        }

        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_FOLLOWLOCATION, @as(c_long, if (req.follow_redirects) 1 else 0));
        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_VERBOSE, @as(c_long, if (req.verbose) 1 else 0));
        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_NOPROGRESS, @as(c_long, 1));

        // Write callback
        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_WRITEFUNCTION, writeCallback);
        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_WRITEDATA, &write_ctx);

        const start = std.time.milliTimestamp();
        const rc = c.curl_easy_perform(self.handle);
        const duration_ms = std.time.milliTimestamp() - start;

        if (rc != c.CURLE_OK) {
            if (rc == c.CURLE_ABORTED_BY_CALLBACK) return error.StreamAborted;
            return error.CurlError;
        }

        var status: c_long = 0;
        _ = c.curl_easy_getinfo(self.handle, c.CURLINFO_RESPONSE_CODE, &status);

        return Response{
            .status_code = @intCast(status),
            .body = try write_ctx.buf.toOwnedSlice(),
            .duration_ms = duration_ms,
        };
    }

    // ------------------------------------------------------------------
    // Streaming request (SSE)
    // ------------------------------------------------------------------

    /// Perform a streaming HTTP request.  `callback` is called for each line
    /// received, including empty lines (SSE event separators).
    /// Returns the HTTP status code; body is consumed by the callback.
    pub fn streamRequest(
        self: *HttpClient,
        allocator: std.mem.Allocator,
        req: Request,
        callback: StreamCallback,
        userdata: ?*anyopaque,
    ) (HttpError || std.mem.Allocator.Error)!u32 {
        c.curl_easy_reset(self.handle);

        var stream_ctx = StreamCtx{
            .callback = callback,
            .userdata = userdata,
            .abort_requested = false,
            .partial = std.ArrayList(u8).init(allocator),
        };
        defer stream_ctx.partial.deinit();

        const url_z = try std.fmt.allocPrintZ(allocator, "{s}", .{req.url});
        defer allocator.free(url_z);

        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_URL, url_z.ptr);

        try setMethod(self.handle, req.method, req.body);

        var slist: ?*c.curl_slist = null;
        defer if (slist) |sl| c.curl_slist_free_all(sl);
        for (req.headers) |h| {
            const header_str = try std.fmt.allocPrintZ(allocator, "{s}: {s}", .{ h.name, h.value });
            defer allocator.free(header_str);
            slist = c.curl_slist_append(slist, header_str.ptr);
        }
        if (slist) |sl| _ = c.curl_easy_setopt(self.handle, c.CURLOPT_HTTPHEADER, sl);

        if (req.connect_timeout_ms > 0) {
            _ = c.curl_easy_setopt(self.handle, c.CURLOPT_CONNECTTIMEOUT_MS, @as(c_long, req.connect_timeout_ms));
        }
        if (req.total_timeout_ms > 0) {
            _ = c.curl_easy_setopt(self.handle, c.CURLOPT_TIMEOUT_MS, @as(c_long, req.total_timeout_ms));
        }

        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_FOLLOWLOCATION, @as(c_long, if (req.follow_redirects) 1 else 0));
        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_VERBOSE, @as(c_long, if (req.verbose) 1 else 0));
        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_NOPROGRESS, @as(c_long, 1));

        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_WRITEFUNCTION, streamWriteCallback);
        _ = c.curl_easy_setopt(self.handle, c.CURLOPT_WRITEDATA, &stream_ctx);

        const rc = c.curl_easy_perform(self.handle);

        if (stream_ctx.abort_requested) return error.StreamAborted;
        if (rc != c.CURLE_OK) {
            if (rc == c.CURLE_ABORTED_BY_CALLBACK) return error.StreamAborted;
            return error.CurlError;
        }

        var status: c_long = 0;
        _ = c.curl_easy_getinfo(self.handle, c.CURLINFO_RESPONSE_CODE, &status);
        return @intCast(status);
    }
};

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

fn setMethod(handle: *c.CURL, method: Method, body: ?[]const u8) HttpError!void {
    switch (method) {
        .GET => {
            _ = c.curl_easy_setopt(handle, c.CURLOPT_HTTPGET, @as(c_long, 1));
        },
        .POST => {
            _ = c.curl_easy_setopt(handle, c.CURLOPT_POST, @as(c_long, 1));
            if (body) |b| {
                _ = c.curl_easy_setopt(handle, c.CURLOPT_POSTFIELDS, b.ptr);
                _ = c.curl_easy_setopt(handle, c.CURLOPT_POSTFIELDSIZE, @as(c_long, @intCast(b.len)));
            } else {
                _ = c.curl_easy_setopt(handle, c.CURLOPT_POSTFIELDSIZE, @as(c_long, 0));
            }
        },
        .PUT => {
            _ = c.curl_easy_setopt(handle, c.CURLOPT_CUSTOMREQUEST, "PUT");
            if (body) |b| {
                _ = c.curl_easy_setopt(handle, c.CURLOPT_POSTFIELDS, b.ptr);
                _ = c.curl_easy_setopt(handle, c.CURLOPT_POSTFIELDSIZE, @as(c_long, @intCast(b.len)));
            }
        },
        .DELETE => {
            _ = c.curl_easy_setopt(handle, c.CURLOPT_CUSTOMREQUEST, "DELETE");
        },
        .PATCH => {
            _ = c.curl_easy_setopt(handle, c.CURLOPT_CUSTOMREQUEST, "PATCH");
            if (body) |b| {
                _ = c.curl_easy_setopt(handle, c.CURLOPT_POSTFIELDS, b.ptr);
                _ = c.curl_easy_setopt(handle, c.CURLOPT_POSTFIELDSIZE, @as(c_long, @intCast(b.len)));
            }
        },
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "HttpClient init and deinit" {
    // Requires libcurl available at link time; just test the construction path.
    // If curl is not present this will fail at link time, not at test time.
    var client = try HttpClient.init();
    defer client.deinit();
    // If we got here curl is available.
    try std.testing.expect(true);
}

test "Request default values" {
    const req = Request{ .url = "https://example.com" };
    try std.testing.expect(req.method == .POST);
    try std.testing.expectEqual(@as(u32, 30_000), req.connect_timeout_ms);
    try std.testing.expectEqual(@as(u32, 300_000), req.total_timeout_ms);
    try std.testing.expect(!req.follow_redirects);
    try std.testing.expect(!req.verbose);
    try std.testing.expect(req.body == null);
    try std.testing.expectEqual(@as(usize, 0), req.headers.len);
}

test "Header struct" {
    const h = Header{ .name = "Authorization", .value = "Bearer token123" };
    try std.testing.expectEqualStrings("Authorization", h.name);
    try std.testing.expectEqualStrings("Bearer token123", h.value);
}

test "Method enum values" {
    const methods = [_]Method{ .GET, .POST, .PUT, .DELETE, .PATCH };
    try std.testing.expectEqual(@as(usize, 5), methods.len);
}
